crates/librqbit/src/blocklist.rs

```diff
@@ -0,0 +1,314 @@
+use anyhow::Result;
+use async_compression::tokio::bufread::GzipDecoder;
+use futures::TryStreamExt;
+use intervaltree::IntervalTree;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::pin::Pin;
+use std::str::FromStr;
+use tokio::io::AsyncRead;
+use tokio::{io::AsyncBufReadExt, io::BufReader};
+use tokio_util::io::StreamReader;
+use tracing::{debug, info, trace};
+
+pub struct Blocklist {
+    // ipv4 and ipv6 do not overlap
+    // see: https://www.rfc-editor.org/rfc/rfc4291#section-2.5.5
+    blocked_ranges: IntervalTree<IpAddr, ()>,
+}
+
+impl Blocklist {
+    pub fn empty() -> Self {
+        return Self::new(std::iter::empty());
+    }
+
+    pub fn new(ip_ranges: impl IntoIterator<Item = std::ops::Range<IpAddr>>) -> Self {
+        Self {
+            blocked_ranges: IntervalTree::from_iter(ip_ranges.into_iter().map(|r| (r, ()))),
+        }
+    }
+
+    pub async fn load_from_url(url: &str) -> Result<Self> {
+        let response = reqwest::get(url).await.map_err(|e| anyhow::anyhow!(e))?;
```

Comment:
you can probably just use .context() here and below.
Also anyhow::bail! is handier than "return Err(anyhow::anyhow!"